tests/cpp_unit_tests/test_math.cpp

```diff
@@ -0,0 +1,364 @@
+#include <boost/math/special_functions/legendre.hpp>
+#include <catch2/catch_test_macros.hpp>
+#include <catch2/matchers/catch_matchers_floating_point.hpp>
+#include <catch2/matchers/catch_matchers_vector.hpp>
+
+#include "openmc/math_functions.h"
+#include "openmc/random_dist.h"
+#include "openmc/wmp.h"
+
+TEST_CASE("Test t_percentile")
+{
+  // Permutations include 1 DoF, 2 DoF, and > 2 DoF
+  // We will test 5 p-values at 3-DoF values
+  std::vector<double> test_ps = {0.02, 0.4, 0.5, 0.6, 0.98};
+  std::vector<int> test_dfs = {1, 2, 5};
+
+  // The reference solutions come from boost::math
+  for (int df : test_dfs) {
+    std::vector<double> ref_ts;
+    std::vector<double> test_ts;
+
+    for (double p : test_ps) {
+      double ref_t = boost::math::quantile(boost::math::students_t(df), p);
+      double test_t = openmc::t_percentile(p, df);
+
+      ref_ts.push_back(ref_t);
+      test_ts.push_back(test_t);
+    }
+
+    // The 5 DoF approximation in openmc.lib.math.t_percentile is off by up to
+    // 8e-3 from the scipy solution, so test that one separately with looser
+    // tolerance
+    double tolerance = (df > 2) ? 1e-2 : 1e-6;
+
+    REQUIRE_THAT(ref_ts, Catch::Matchers::Approx(test_ts).epsilon(tolerance));
+  }
+}
+
+TEST_CASE("Test calc_pn")
+{
+  int max_order = 10;
+  std::vector<double> test_xs = {-1.0, -0.5, 0.0, 0.5, 1.0};
+
+  std::vector<std::vector<double>> ref_vals;
+  std::vector<std::vector<double>> test_vals;
+  for (double x : test_xs) {
+    // Reference solutions from boost::math
+    std::vector<double> ref_val;
+
+    for (int n = 0; n <= max_order; ++n) {
+      ref_val.push_back(boost::math::legendre_p(n, x));
+    }
+    ref_vals.push_back(ref_val);
+
+    std::vector<double> test_val(max_order + 1);
+    openmc::calc_pn_c(max_order, x, test_val.data());
+    test_vals.push_back(test_val);
+  }
+
+  for (int i = 0; i < ref_vals.size(); i++) {
+    REQUIRE_THAT(ref_vals[0], Catch::Matchers::Approx(test_vals[0]));
+  }
+}
+
+TEST_CASE("Test evaluate_legendre")
+{
+  int max_order = 10;
+  std::vector<double> test_xs = {-1.0, -0.5, 0.0, 0.5, 1.0};
+
+  std::vector<double> ref_vals;
+
+  for (double x : test_xs) {
+    double sum = 0.0;
+
+    // Coefficients are set to 1, but will incorporate the (2l+1)/2 norm factor
+    // for the reference solution
+    for (int l = 0; l <= max_order; ++l) {
+      double coeff = 0.5 * (2 * l + 1);
+      sum += coeff * boost::math::legendre_p(l, x);
+    }
+
+    ref_vals.push_back(sum);
+  }
+
+  // Set the coefficients back to 1s for the test values since
+  // evaluate legendre incorporates the (2l+1)/2 term on its own
+  std::vector<double> test_coeffs(max_order + 1, 1.0);
+
+  std::vector<double> test_vals;
+  for (double x : test_xs) {
+    test_vals.push_back(
+      openmc::evaluate_legendre(test_coeffs.size() - 1, test_coeffs.data(), x));
+  }
+
+  REQUIRE_THAT(ref_vals, Catch::Matchers::Approx(test_vals));
+}
+
+double coeff(int n, int m)
```

Comment:
Can you rename this to be more descriptive than coeff?

Reply:
ok, no problem, I will change reference using scipy, and rename the internal function name